File: 1825.finding-mk-average.cpp
/*
 * @lc app=leetcode id=1825 lang=cpp
 *
 * [1825] Finding MK Average
 */
#include <bits/stdc++.h>

using std::multiset;
using std::vector;

// @lc code=start
class MKAverage {
    size_t m, k, sz, p;
    long sum;
    vector<int> circular;  // can also use a queue

    // std::multiset is implemented with balanced tree (red-black tree, AVL tree, etc.).
    // Rebalance the following sets when adding or removing numbers,
    // such that the sizes of left and right are both k.
    multiset<int> left, mid, right;

    // Rebalance helper functions
    void add(int x) {
        left.insert(x);
        if (left.size() > k) {
            auto it = std::prev(left.end());
            mid.insert(*it);
            sum += *it;
            left.erase(it);
        }
        if (mid.size() > sz) {
            auto it = std::prev(mid.end());
            right.insert(*it);
            sum -= *it;
            mid.erase(it);
        }
    }

    void remove(int x) {
        if (x <= *left.rbegin()) {
            left.erase(left.find(x));
        } else if (x <= *mid.rbegin()) {
            auto it = mid.find(x);
            sum -= *it;
            mid.erase(it);
        } else {
            right.erase(right.find(x));
        }

        if (left.size() < k) {
            auto mid_min_it = mid.begin();
            left.insert(*mid_min_it);
            sum -= *mid_min_it;
            mid.erase(mid_min_it);
        }
        if (mid.size() < sz) {
            auto right_min_it = right.begin();
            mid.insert(*right_min_it);
            sum += *right_min_it;
            right.erase(right_min_it);
        }
    }

public:
    MKAverage(int m, int k) : m(m), k(k), sz(m - (k << 1)), p(0), sum(0), circular(m) {}
    
    // T: O(log(m))
    void addElement(int num) {
        if (p >= m) {
            remove(circular[p % m]);
        }
        add(num);
        circular[p++ % m] = num;
    }
    
    // T: O(1)
    int calculateMKAverage() {
        return p >= m ? sum / sz : -1;
    }
};

/**
 * Your MKAverage object will be instantiated and called as such:
 * MKAverage* obj = new MKAverage(m, k);
 * obj->addElement(num);
 * int param_2 = obj->calculateMKAverage();
 */
// @lc code=end

File: 1825.finding-mk-average.java
/*
 * @lc app=leetcode id=1825 lang=java
 *
 * [1825] Finding MK Average
 */
import java.util.*;

// @lc code=start
class MKAverage {
    private int m, k, sz;
    private Queue<Integer> q;
    private long sum;
    private TreeMap<Integer, Integer> left, mid, right;
    private int leftSz, midSz, rightSz;

    public MKAverage(int m, int k) {
        this.m = m;
        this.k = k;
        sz = m - (k << 1);
        q = new ArrayDeque<>();
        left = new TreeMap<>();
        mid = new TreeMap<>();
        right = new TreeMap<>();
    }
    
    // T: O(log(m))
    public void addElement(int num) {
        if (q.size() == m) {
            remove(q.poll());
        }
        add(num);
        q.offer(num);
    }
    
    // T: O(1)
    public int calculateMKAverage() {
        return q.size() == m ? (int)(sum / sz) : -1;
    }

    private void add(int x) {
        left.put(x, left.getOrDefault(x, 0) + 1);
        ++leftSz;
        if (leftSz > k) {
            int leftMax = left.lastKey();
            addToMap(mid, leftMax);
            ++midSz;
            sum += leftMax;
            removeFromMap(left, leftMax);
            --leftSz;
        }
        if (midSz > sz) {
            int midMax = mid.lastKey();
            addToMap(right, midMax);
            ++rightSz;
            sum -= midMax;
            removeFromMap(mid, midMax);
            --midSz;
        }
    }

    private void remove(int x) {
        if (left.containsKey(x)) {
            removeFromMap(left, x);
            --leftSz;
        } else if (mid.containsKey(x)) {
            sum -= x;
            removeFromMap(mid, x);
            --midSz;
        } else {
            removeFromMap(right, x);
            --rightSz;
        }

        if (leftSz < k) {
            int midMin = mid.firstKey();
            addToMap(left, midMin);
            ++leftSz;
            sum -= midMin;
            removeFromMap(mid, midMin);
            --midSz;
        }
        if (midSz < sz) {
            int rightMin = right.firstKey();
            addToMap(mid, rightMin);
            ++midSz;
            sum += rightMin;
            removeFromMap(right, rightMin);
            --rightSz;
        }
    }

    private static void addToMap(TreeMap<Integer, Integer> mp, int x) {
        mp.put(x, mp.getOrDefault(x, 0) + 1);
    }

    private static void removeFromMap(TreeMap<Integer, Integer> mp, int x) {
        mp.put(x, mp.get(x) - 1);
        if (mp.get(x) == 0) {
            mp.remove(x);
        }
    }
}

/**
 * Your MKAverage object will be instantiated and called as such:
 * MKAverage obj = new MKAverage(m, k);
 * obj.addElement(num);
 * int param_2 = obj.calculateMKAverage();
 */
// @lc code=end

